Expose underlying unix file descriptor in SafeMemoryMappedFileHandle

dannyfriar opened the pull request:
See https://github.com/dotnet/runtime/issues/52854

Done in a similar way to [SafePipeHandle](https://github.com/dotnet/runtime/blob/01b7e73cd378145264a7cb7a09365b41ed42b240/src/libraries/System.IO.Pipes/src/Microsoft/Win32/SafeHandles/SafePipeHandle.Unix.cs#L106).

Commit:
expose unix file descriptor safe memory map

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=53538) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=53538) before we can accept your contribution.<br/><br/>:x: dannyfriar [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=53538)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=53538) it.</sub>

stephentoub commented on src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs:
```diff
@@ -59,9 +57,22 @@ public sealed partial class SafeMemoryMappedFileHandle : SafeHandleZeroOrMinusOn
+            IntPtr handlePtr;
+
+            if (fileStream != null) {
+                _fileStreamHandle = fileStream.SafeFileHandle;
```
This should really be stored only after calling DangerousAddRef.  The issue is that if DangerousAddRef throws, the finalizer for this object will still be invoked, and if it sees _fileStreamHandle as non-null, it'll call DangerousRelease.  Now, I see we do this same ordering in pipes, and in reality it's unlikely to be an actual issue, since the only reason DangerousAddRef should throw is if the handle has already been disposed, in which case an extraneous DangerousRelease on it in the finalizer shouldn't matter functionally.  But, still good hygiene.

stephentoub commented on src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs:
```diff
@@ -59,9 +57,22 @@ public sealed partial class SafeMemoryMappedFileHandle : SafeHandleZeroOrMinusOn
+                handlePtr = _fileStreamHandle.DangerousGetHandle();
+            }
+            else {
+                _fileStreamHandle = null;
```
Why is this needed?

stephentoub commented on src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs:
```diff
@@ -74,7 +85,17 @@ protected override void Dispose(bool disposing)
 
-        protected override bool ReleaseHandle() => true; // Nothing to clean up.  We unlinked immediately after creating the backing store.
+        protected override bool ReleaseHandle() {
+            Debug.Assert(!IsInvalid);
```
This assert isn't valid.  The ctor could throw before reaching the SetHandle call, in which case the handle will still have a value of 0, which is considered invalid.

stephentoub commented on src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs:
```diff
@@ -74,7 +85,17 @@ protected override void Dispose(bool disposing)
+            }
+
+            return true;
+        }
```
Can you please add test validating the desired functionality?

dannyfriar replied:
Sure, I will add

dannyfriar replied:
@stephentoub I've added a test to check it matches the filestream handle

Commit:
store handle after DangerousAddRef

Commit:
add test verifying handle matches filestream handle

stephentoub commented on src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs:
```diff
@@ -59,9 +57,21 @@ public sealed partial class SafeMemoryMappedFileHandle : SafeHandleZeroOrMinusOn
+                _fileStreamHandle = fileStream.SafeFileHandle;
+                handlePtr = _fileStreamHandle.DangerousGetHandle();
+            }
+            else {
```
Nit:
```suggestion
            else
            {
```

stephentoub commented on src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs:
```diff
@@ -74,7 +84,15 @@ protected override void Dispose(bool disposing)
 
-        protected override bool ReleaseHandle() => true; // Nothing to clean up.  We unlinked immediately after creating the backing store.
+        protected override bool ReleaseHandle() {
+            if (_fileStreamHandle != null) {
```
```suggestion
            if (_fileStreamHandle != null)
            {
```

stephentoub commented on src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs:
```diff
@@ -74,7 +84,15 @@ protected override void Dispose(bool disposing)
         }
 
-        protected override bool ReleaseHandle() => true; // Nothing to clean up.  We unlinked immediately after creating the backing store.
+        protected override bool ReleaseHandle() {
```
```suggestion
        protected override bool ReleaseHandle()
        {
```

stephentoub commented on src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs:
```diff
@@ -59,9 +57,21 @@ public sealed partial class SafeMemoryMappedFileHandle : SafeHandleZeroOrMinusOn
+            if (fileStream != null)
+            {
+                bool ignored = false;
+                fileStream.SafeFileHandle.DangerousAddRef(ref ignored);
+
+                _fileStreamHandle = fileStream.SafeFileHandle;
+                handlePtr = _fileStreamHandle.DangerousGetHandle();
```
```suggestion
                SafeFileHandle handle = fileStream.SafeFileHandle;
                handle.DangerousAddRef(ref ignored);
                _fileStreamHandle = handle;
                handlePtr = handle.DangerousGetHandle();
```

stephentoub commented on src/libraries/System.IO.MemoryMappedFiles/tests/MemoryMappedFile.CreateFromFile.Tests.cs:
```diff
@@ -949,5 +949,21 @@ public void OpenCharacterDeviceAsFile(MemoryMappedFileAccess access)
+        public void MapHandleMatchesFileStreamHandle()
+        {
+            using (FileStream fs = File.OpenWrite(GetTestFilePath())) {
+                MemoryMappedFile mmf = MemoryMappedFile.CreateFromFile(
```
This should be disposed as well.

stephentoub commented on src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs:
```diff
@@ -11,15 +11,13 @@ namespace Microsoft.Win32.SafeHandles
         /// </summary>
-        internal readonly FileStream? _fileStream;
+        private readonly FileStream? _fileStream;
+        internal SafeFileHandle? _fileStreamHandle;
```
```suggestion
        /// <summary>The FileStream's handle, cached to avoid repeated accesses to FileStream.SafeFileHandle that could, in theory, change.</summary>
        internal SafeFileHandle? _fileStreamHandle;
```

stephentoub reviewed (approved):
Left some comments to be addressed, but otherwise looks good. Thanks.

dannyfriar committed:
Update src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs

Co-authored-by: Stephen Toub <stoub@microsoft.com>

dannyfriar committed:
Update src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs

Co-authored-by: Stephen Toub <stoub@microsoft.com>

dannyfriar committed:
Update src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs

Co-authored-by: Stephen Toub <stoub@microsoft.com>

dannyfriar committed:
Update src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs

Co-authored-by: Stephen Toub <stoub@microsoft.com>

dannyfriar committed:
Update src/libraries/System.IO.MemoryMappedFiles/tests/MemoryMappedFile.CreateFromFile.Tests.cs

Co-authored-by: Stephen Toub <stoub@microsoft.com>

dannyfriar committed:
Dispose MemoryMappedFile in MapHandleMatchesFileStreamHandle test

dannyfriar committed:
Update src/libraries/System.IO.MemoryMappedFiles/src/Microsoft/Win32/SafeMemoryMappedFileHandle.Unix.cs

Co-authored-by: Stephen Toub <stoub@microsoft.com>

dannyfriar committed:
Update MemoryMappedFile.CreateFromFile.Tests.cs

dannyfriar committed:
Move mmf in test to using block

dannyfriar committed:
Fix test

dannyfriar commented:
> Left some comments to be addressed, but otherwise looks good. Thanks.

Thanks @stephentoub. I've addressed your comments (the test failures are unrelated I think).
